plugin/src/main/java/at/helpch/chatchat/command/SwitchChannelCommand.java

```diff
@@ -20,21 +27,59 @@ public SwitchChannelCommand(@NotNull final ChatChatPlugin plugin, @NotNull final
     }
 
     @Default
-    public void switchChannel(final ChatUser user) {
-
+    public void switchChannel(final ChatUser user, @Join @Optional @NotNull final String message) {
         final var channels = plugin.configManager().channels().channels();
         final var channel = channels.values()
-                .stream()
-                .filter(value -> value.commandName().equals(command))
-                .findAny()
-                .get(); // this should probably only ever throw if the person has changed command names without restarting
+            .stream()
+            .filter(value -> value.commandName().equals(command))
+            .findAny()
+            .get(); // this should probably only ever throw if the person has changed command names without restarting
 
         if (!user.canUse(channel)) {
             user.sendMessage(Component.text("You don't have permission to use this channel!", NamedTextColor.RED));
             return;
         }
 
+        if (message.isEmpty()) {
+            user.channel(channel);
+            user.sendMessage(Component.text("You have switched to the " + command + " channel!", NamedTextColor.GREEN));
+            return;
+        }
+
+        if (StringUtils.containsIllegalChars(message) && !user.player().hasPermission(UTF_PERMISSION)) {
+            user.sendMessage(Component.text("You can't use special characters in chat!", NamedTextColor.RED));
+            return;
+        }
+
+        final var format = FormatUtils.findFormat(user.player(), plugin.configManager().formats());
+
+        final var audience = plugin.usersHolder().users()
+            .stream()
+            .filter(otherUser -> otherUser.canSee(channel)) // get everyone who can see this channel
+            .collect(Audience.toAudience());
+
+        final var chatEvent = new ChatChatEvent(
+            false,
+            user,
+            audience,
+            format,
+            Component.text(message),
+            channel
+        );
+
+        plugin.getServer().getPluginManager().callEvent(chatEvent);
+
+        if (chatEvent.isCancelled()) {
+            return;
+        }
+
+        final var oldChannel = user.channel();
         user.channel(channel);
-        user.sendMessage(Component.text("You have switched to the " + command + " channel!", NamedTextColor.GREEN));
+        chatEvent.recipients().sendMessage(FormatUtils.parseFormat(
+            chatEvent.format(),
+            user.player(),
+            chatEvent.message()
+        ));
+        user.channel(oldChannel);
```

Comment:
This should be extracted into its own method alongside the ChatListener to DRY up the code